File: oapp/services/restaurant/internal/utils/validation.go
package utils

import (
	"fmt"
	"reflect"
	"regexp"

	"github.com/google/uuid"
)

var (
	// PhoneRegex validates phone numbers with international format support
	PhoneRegex = regexp.MustCompile(`^[\d\s\-\+\(\)]+$`)
)

// ValidatePhone checks if a phone number matches the basic pattern
func ValidatePhone(phone string) error {
	if phone == "" {
		return fmt.Errorf("phone number is required")
	}
	if !PhoneRegex.MatchString(phone) {
		return fmt.Errorf("phone number contains invalid characters")
	}
	return nil
}

// ValidateLatitude checks if latitude is within valid range [-90, 90]
func ValidateLatitude(lat float64) error {
	if lat < -90 || lat > 90 {
		return fmt.Errorf("latitude must be between -90 and 90")
	}
	return nil
}

// ValidateLongitude checks if longitude is within valid range [-180, 180]
func ValidateLongitude(lng float64) error {
	if lng < -180 || lng > 180 {
		return fmt.Errorf("longitude must be between -180 and 180")
	}
	return nil
}

// IsEmptyUpdateRequest checks if all pointer fields in an update request are nil
// Returns true if the request would be a no-op (all fields nil)
func IsEmptyUpdateRequest(fields ...*string) bool {
	for _, field := range fields {
		if field != nil {
			return false
		}
	}
	return true
}

// HasAtLeastOneField checks if at least one field is non-nil
// Uses reflection to check the actual pointer value, not the interface
func HasAtLeastOneField(fields ...interface{}) bool {
	for _, field := range fields {
		if field == nil {
			continue
		}
		// Use reflection to check if the pointer is non-nil
		v := reflect.ValueOf(field)
		if v.Kind() == reflect.Ptr && !v.IsNil() {
			return true
		}
		// For non-pointer types, if field != nil, it has a value
		if v.Kind() != reflect.Ptr {
			return true
		}
	}
	return false
}

// IsValidUUID checks if a string is a valid UUID
func IsValidUUID(id string) bool {
	_, err := uuid.Parse(id)
	return err == nil
}

File: admin-portal/src/utils/validation.js
/**
 * Validation utilities for form fields
 * Matches backend validation patterns
 */

// Phone number validation: allows digits, spaces, dashes, plus signs, parentheses
const phoneRegex = /^[\d\s\-\+\(\)]+$/;

// URL validation: must start with http:// or https://
const urlRegex = /^https?:\/\/.+/i;

/**
 * Validate phone number format
 * @param {string} phone - Phone number to validate
 * @returns {string|true} - Error message or true if valid
 */
export const validatePhone = (phone) => {
  if (!phone || phone.trim() === '') {
    return 'Phone number is required';
  }

  if (!phoneRegex.test(phone)) {
    return 'Phone number contains invalid characters';
  }

  // Check if phone has at least one digit
  if (!/\d/.test(phone)) {
    return 'Phone number must contain at least one digit';
  }

  return true;
};

/**
 * Validate URL format (optional field)
 * @param {string} url - URL to validate
 * @returns {string|true} - Error message or true if valid
 */
export const validateUrl = (url) => {
  // Empty is valid (optional field)
  if (!url || url.trim() === '') {
    return true;
  }

  if (!urlRegex.test(url)) {
    return 'URL must start with http:// or https://';
  }

  return true;
};

/**
 * Validate latitude range
 * @param {number} lat - Latitude value
 * @returns {string|true} - Error message or true if valid
 */
export const validateLatitude = (lat) => {
  // Empty is valid (optional field)
  if (lat === undefined || lat === null || lat === '') {
    return true;
  }

  const numLat = parseFloat(lat);

  if (isNaN(numLat)) {
    return 'Latitude must be a number';
  }

  if (numLat < -90 || numLat > 90) {
    return 'Latitude must be between -90 and 90';
  }

  return true;
};

/**
 * Validate longitude range
 * @param {number} lng - Longitude value
 * @returns {string|true} - Error message or true if valid
 */
export const validateLongitude = (lng) => {
  // Empty is valid (optional field)
  if (lng === undefined || lng === null || lng === '') {
    return true;
  }

  const numLng = parseFloat(lng);

  if (isNaN(numLng)) {
    return 'Longitude must be a number';
  }

  if (numLng < -180 || numLng > 180) {
    return 'Longitude must be between -180 and 180';
  }

  return true;
};

/**
 * Validate field length
 * @param {string} value - Value to validate
 * @param {number} maxLength - Maximum allowed length
 * @param {string} fieldName - Name of field for error message
 * @returns {string|true} - Error message or true if valid
 */
export const validateLength = (value, maxLength, fieldName = 'Field') => {
  if (!value) {
    return true; // Let required validation handle empty values
  }

  if (value.length > maxLength) {
    return `${fieldName} must not exceed ${maxLength} characters`;
  }

  return true;
};

/**
 * Get store initials for placeholder logo
 * @param {string} storeName - Store name
 * @returns {string} - Initials (max 2 characters)
 */
export const getStoreInitials = (storeName) => {
  if (!storeName) return '?';

  const words = storeName.trim().split(/\s+/);
  if (words.length === 1) {
    return words[0].substring(0, 2).toUpperCase();
  }

  return (words[0][0] + words[1][0]).toUpperCase();
};
